CI: split formal and build to fix posting comments

GeorgeSapkin opened the pull request:
Separate formal from build so the event trigger can be set to pull_request_target and fix posting comments. This event allows workflow to do things like label or comment on pull requests from forks, but it's not recommended for build jobs due to security implications.

Switch build to workflow_dispatch and trigger it from formal when it succeeds. Dispatching a workflow requires a fine-grained token, exposed through `FORMAL_TOKEN` secret. Since the workflow is no longer running in a PR context, it needs the PR info passed to it to manually make the appearance of being tied to said PR by setting run-name, and passing back individual matrix step statuses manually to the PR with all the correct info. To that end re-introduce the full build workflow that was moved to actions for easier testing with the expectation of moving it back to actions once it's fully tested.

Build workflow can now be triggered from the UI, but it's limited to repo members.

The end result looks mostly the same, but now with the added bonus of some more statuses being visible in the PR like both build and test.

<img width="870" height="406" alt="Screenshot From 2025-12-06 05-25-54" src="https://github.com/user-attachments/assets/c59783d5-a371-4665-8bbe-efff5e67f323" />

The downsides:

- Requires a fine-grained access token from some bot user and build will appear as being triggered by this bot user.
- Current reusable build workflow (the one in the actions repo) needs be modified to do all the status magic to pretend it's running the PR context.
- Manually canceling the run will have hanging statuses in the PR check UI.
- Build can be triggered from the UI, although this can be filter on the allowed users of course:

<img width="394" height="475" alt="image" src="https://github.com/user-attachments/assets/65ce0994-bb23-4a0e-8397-bb5090e75bf0" />

Fixes: 7658669 ("multi-arch-test-build: post formal summaries to PR")
Link: https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows#pull_request_target
Link: https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows#workflow_dispatch
Link: https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/
Link: https://securitylab.github.com/resources/github-actions-new-patterns-and-mitigations/

Requires:
- [ ] `FORMAL_TOKEN` with `actions: write` permission defined in secrets.

<img width="804" height="105" alt="image" src="https://github.com/user-attachments/assets/30cd3935-4f83-41db-8254-df4dbaae972c" />

Related:
- #27790
- #27943

cc: @BKPepe, @hnyman, @systemcrash

hnyman commented:
This starts to sound pretty complicated.  I wonder about the maintainability in future...

GeorgeSapkin committed:
ci: split formal and build to fix posting comments

Separate formal from build so the event trigger can be set to
pull_request_target and fix posting comments. This event allows
workflow to do things like label or comment on pull requests from forks,
but it's not recommended for build jobs due to security implications.

Switch build to workflow_dispatch and trigger it from formal when it
succeeds. Dispatching a workflow requires a fine-grained token, exposed
through FORMAL_TOKEN secret. Since the workflow is no longer running in
a PR context, it needs the PR info passed to it to manually make the
appearance of being tied to said PR by setting run-name, and passing
back individual matrix step statuses manually to the PR with all the
correct info. To that end re-introduce the full build workflow that was
moved to actions for easier testing with the expectation of moving it
back to actions once it's fully tested.

FORMAL_TOKEN requires actions: write permission.

Build workflow can now be triggered from the UI, but it's limited to
repo members.

Fixes: 7658669 ("multi-arch-test-build: post formal summaries to PR")
Link: https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows#pull_request_target
Link: https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows#workflow_dispatch
Link: https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/
Link: https://securitylab.github.com/resources/github-actions-new-patterns-and-mitigations/
Signed-off-by: George Sapkin <george@sapk.in>

hnyman commented:
> I think we should ping somebody with access to the @openwrt-bot to create the necessary token.

Please do, because right now the currently semi-broken fancy CI is practically unusable, as it prevents understanding if the PRs actually fail. The new fancy notifications cause the CI run to stop also for warnings, and then the build test is not done at all.

If you can't get this fixed/implemented soon, we should revert to the previous version that actually worked.

robimarko commented:
I am pinging people to see who can actually acess the openwrt-bot account so the token can be generated
